src/platform/ESP32/ESP32Utils.cpp

```diff
@@ -237,3 +238,70 @@ bool ESP32Utils::HasIPv6LinkLocalAddress(tcpip_adapter_if_t intfId)
     ip6_addr_t unused;
     return tcpip_adapter_get_ip6_linklocal(intfId, &unused) == ESP_OK;
 }
+
+CHIP_ERROR ESP32Utils::GetWiFiStationProvision(Internal::DeviceNetworkInfo & netInfo, bool includeCredentials)
+{
+    CHIP_ERROR err = CHIP_NO_ERROR;
+    wifi_config_t stationConfig;
+
+    err = esp_wifi_get_config(ESP_IF_WIFI_STA, &stationConfig);
+    SuccessOrExit(err);
+
+    VerifyOrExit(stationConfig.sta.ssid[0] != 0, err = CHIP_ERROR_INCORRECT_STATE);
+
+    netInfo.NetworkId              = kWiFiStationNetworkId;
+    netInfo.FieldPresent.NetworkId = true;
+    memcpy(netInfo.WiFiSSID, stationConfig.sta.ssid, min(strlen((char *) stationConfig.sta.ssid) + 1, sizeof(netInfo.WiFiSSID)));
+
+    if (includeCredentials)
+    {
+        netInfo.WiFiKeyLen = min(strlen((char *) stationConfig.sta.password), sizeof(netInfo.WiFiKey));
+        memcpy(netInfo.WiFiKey, stationConfig.sta.password, netInfo.WiFiKeyLen);
+    }
+
+exit:
+    return err;
+}
+
+CHIP_ERROR ESP32Utils::SetWiFiStationProvision(const Internal::DeviceNetworkInfo & netInfo)
+{
+    CHIP_ERROR err = CHIP_NO_ERROR;
+    wifi_config_t wifiConfig;
+
+    // Ensure that ESP station mode is enabled.  This is required before esp_wifi_set_config(ESP_IF_WIFI_STA,...)
+    // can be called.
+    err = ESP32Utils::EnableStationMode();
+    SuccessOrExit(err);
+
+    // Initialize an ESP wifi_config_t structure based on the new provision information.
+    memset(&wifiConfig, 0, sizeof(wifiConfig));
+    memcpy(wifiConfig.sta.ssid, netInfo.WiFiSSID, min(strlen(netInfo.WiFiSSID) + 1, sizeof(wifiConfig.sta.ssid)));
```

Comment:
should we ensure null terminator here and on the password?

Reply:
It seems to be documented as not having a null terminator (and in weave as well).

Reply:
Yes this is actually Weave code in CHIP